src/status_im/chat/models/message.cljs

```diff
@@ -36,16 +37,64 @@
     (not current-chat?) (assoc :appearing? false)
     (emoji-only-content? content) (assoc :content-type constants/content-type-emoji)))
 
+(defn re-index-message-groups
+  "Relative datemarks of message groups can get obsolete with passing time,
+  this function re-indexes them for given chat"
+  [chat-id {:keys [db]}]
+  (let [chat-messages (get-in db [:chats chat-id :messages])]
+    {:db (update-in db
+                    [:chats chat-id :message-groups]
+                    (partial reduce-kv (fn [groups datemark message-refs]
+                                         (let [new-datemark (->> message-refs
+                                                                 first
+                                                                 :message-id
+                                                                 (get chat-messages)
+                                                                 :timestamp
+                                                                 time/day-relative)]
+                                           (if (= datemark new-datemark)
+                                             ;; nothing to re-index
+                                             (assoc groups datemark message-refs)
+                                             ;; relative datemark shifted, reindex
+                                             (assoc groups new-datemark message-refs))))
+                             {}))}))
+
+(defn- sort-references
+  [messages message-references]
+  (sort-by (juxt (comp :clock-value (partial get messages) :message-id)
+                 :message-id)
+           message-references))
+
+(defn group-messages
+  "Takes chat-id, new messages + cofx and properly groups them into the `:message-groups` index in db"
+  [chat-id messages {:keys [db]}]
+  {:db (reduce (fn [db [datemark grouped-messages]]
+                 (update-in db [:chats chat-id :message-groups datemark]
+                            (fn [message-references]
```

Comment:
Would be great to extract this function as well (again just my opinion)

Reply:
Good idea